### ISSUE TITLE
Feature Request: Allow encoding/other operations on fuzzing input

### ISSUE THREAD

**Issue by Ice3man543** [ffuf/ffuf#149]
It would be great if ffuf could allow the user to encode/perform other operations on the fuzzing input, just like burp suite intruder.

This could be used for Basic authorization bruteforcing for instance. The user could pass two lists and do base64 encode on `fuzz_input_1:fuzz_input_2`

**Comment by bsysop** [comment]
Interesting, if im not wrong, you can use an external mutator, but of course you should do that mutation by your self!

**Comment by Ice3man543** [comment]
I think we could use a template engine for this type of substitution. Not sure of the performance impact though. I am afraid adding it will introduce too much complexity.

**Comment by bsysop** [comment]
CPU/DDR will be nervous :)

What about improve integration and develop external tools in go?

**Comment by Ice3man543** [comment]
@bsysop that could be done, however I don't want to have to write code to do this kind of thing and recompile and run it. 

We could add support for something like js engine that would be optional. It would only be run if the user desired it and perform operations as needed. I really wouldn't mind that much if during creds bruteforce CPU usage rises a bit.

Here is a reference engine - 
[https://github.com/robertkrimen/otto](https://github.com/robertkrimen/otto).

**Comment by bsysop** [comment]
That is really interesting, i really trust GO in performance question, i have wrote programs in many languages to test performance in recon/exploring bugs, no chance to anothers.

I think ffuf main concept is speed, because that he come so popular, that implementation with JS would be great, but also can be done with external tool that integrate with JS.

ffuf
 - ffuf_modules
    - otto_js_parser..
       - your_script_a
       - your_script_b

Just ideias

**Comment by joohoi** [comment]
Currently, `ffuf` is written completely against the Go standard library, with zero external dependencies. For multiple reasons, I prefer trying to keep it like that if not completely necessary.

Furthermore, this functionality can already be achieved with `--input-cmd` by writing a short shell script that reads two wordlists and creates those base64 values for ffuf to use.

The ffuf command would look something like:
```
ffuf --input-cmd 'basic_auth.sh' -u https://example.com/auth -H 'Authorization: Basic FUZZ' ...
```

That said, it would probably make sense to create a library of different payload generators that are designed for the `--input-cmd` in mind. What do you think about this approach?

**Comment by bsysop** [comment]
@joohoi keep simple I think it’s a very important point, against that ffuf would be a framework, complex and doing all things who can be confusing and slow

About “ --input-cmd” I think is perfect, maybe we can do some samples for people, but I have a question about “FUZZ” keyword, can we combine that with another wordlist or customize input-cmd keyword?

If yes I think is complete

**Comment by joohoi** [comment]
Absolutely, the `--input-cmd` can be combined with other wordlists / keywords, and supports custom keyword assignation as well. 

Another way of achieving this is reading the base64 values from stdin, so you can have a script creating the values, printing them out one value per line and just pipe it:
```
./basic_auth.sh | ffuf -w - -u https://example.com/auth -H 'Authorization: Basic FUZZ' ...
```
TL;DR: if you provide `-` as an argument to `-w`, the inputs are read from stdin.

**Comment by bsysop** [comment]
Jesus Christ @joohoi 😍

@Ice3man543 you think this can support what you need? Some encoder that you need now?

**Comment by Ice3man543** [comment]
@joohoi @bsysop I think that'd be good enough for my use case. If you could add sample scripts it'd be much appreciated! But this approach looks good to me!

**Comment by bsysop** [comment]
I could do that a bit later!

@joohoi just send here or wanna share in specific place?

**Comment by joohoi** [comment]
I actually ended up writing one, and creating a new repository for this and other scripts that help in payload generation: [https://github.com/ffuf/ffuf-scripts](https://github.com/ffuf/ffuf-scripts) using the basic authentication payload generating script as an example.

**Comment by Ice3man543** [comment]
@joohoi excellent, that's exactly what i needed. The `ffuf-scripts` repo will be a goldmine i believe. Thanks for working on it. Closing the issue now!

**Comment by D3vil0per** [comment]
> I actually ended up writing one, and creating a new repository for this and other scripts that help in payload generation: https://github.com/ffuf/ffuf-scripts using the basic authentication payload generating script as an example.

Hey @joohoi that script is still maintained? Just one question: could be better if you invert at the end of the script on the nested loop the user and the password in order to test N usernames for one password, instead of 1 username for M passwords? Could be more efficient mostly for cases where the password wordlist is much bigger than username list (that is the most probable case). Thanks

**Comment by HarimbolaSantatra** [comment]
So much information here! I think these info should be included into the wiki!
